jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.cpp

```diff
@@ -118,14 +120,220 @@ ecma_builtin_number_prototype_object_to_string (ecma_value_t this_arg, /**< this
     return ecma_make_throw_obj_completion_value (ecma_new_standard_error (ECMA_ERROR_TYPE));
   }
 
-  if (arguments_list_len == 0)
+  if (arguments_list_len == 0
+      || ecma_number_is_nan (this_arg_number)
+      || ecma_number_is_infinity (this_arg_number)
+      || ecma_number_is_zero (this_arg_number))
   {
     ecma_string_t *ret_str_p = ecma_new_ecma_string_from_number (this_arg_number);
 
     return ecma_make_normal_completion_value (ecma_make_string_value (ret_str_p));
   }
+  else
   {
-    ECMA_BUILTIN_CP_UNIMPLEMENTED (arguments_list_p);
+    const lit_utf8_byte_t digit_chars[36] =
+    {
+      '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+      'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j',
+      'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't',
+      'u', 'v', 'w', 'x', 'y', 'z'
+    };
+
+    ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+    ECMA_OP_TO_NUMBER_TRY_CATCH (arg_num, arguments_list_p[0], ret_value);
+
+    uint32_t radix = ecma_number_to_uint32 (arg_num);
+
+    if (radix < 2 || radix > 36)
+    {
+      ret_value = ecma_make_throw_obj_completion_value (ecma_new_standard_error (ECMA_ERROR_RANGE));
+    }
+    else if (radix == 10)
+    {
+      ecma_string_t *ret_str_p = ecma_new_ecma_string_from_number (this_arg_number);
+
+      ret_value = ecma_make_normal_completion_value (ecma_make_string_value (ret_str_p));
+    }
+    else
+    {
+      const int precision = 53;
+
+      uint64_t digits;
+      int32_t num_digits;
+      int32_t exponent;
+      bool is_negative = false;
+
+      if (ecma_number_is_negative (this_arg_number))
+      {
+        this_arg_number = -this_arg_number;
+        is_negative = true;
+      }
+
+      ecma_number_to_decimal (this_arg_number, &digits, &num_digits, &exponent);
+
+      exponent = exponent - num_digits;
+      bool is_scale_negative = false;
+
+      /* Calculate the scale of the number in the specified radix. */
+      int scale = (int) -floor ((log (10) / log (radix)) * exponent);
+
+      int buff_size;
+
+      if (is_scale_negative)
+      {
+        buff_size = (int) floor ((log (this_arg_number) / log (10))) + 1;
+      }
+      else
+      {
+        buff_size = scale + precision + 2;
+      }
+
+      if (is_negative)
+      {
+        buff_size++;
+      }
+
+      if (scale < 0)
+      {
+        is_scale_negative = true;
+        scale = -scale;
+      }
+
+      /* Normalize the number, so that it is as close to 0 exponent as possible. */
+      for (int i = 0; i < scale; i++)
+      {
+        if (is_scale_negative)
+        {
+          this_arg_number /= (ecma_number_t) radix;
+        }
+        else
+        {
+          this_arg_number *= (ecma_number_t) radix;
+        }
+      }
+
+      uint64_t whole = (uint64_t) this_arg_number;
+      double fraction = this_arg_number - (double) whole;
```

Comment:
This was unintentional. But since you asked, I don't think this improves precision. The original value is already constrained to a float, and because of the continued multiplication the required precision to represent the result would only decrease.